### PR TITLE
Fix duplicate notification profile version numbers on concurrent edits

### PR DESCRIPTION
**Pull request by 3keyroman**
## What changed

Concurrent edits of a notification profile could assign the same version number twice: `editNotificationProfile` computed the next version from the latest existing row without any serialization, and the schema had no uniqueness guarantee on `(notification_profile_uuid, version)`. Duplicate version numbers made version resolution for notifications nondeterministic.

- `editNotificationProfile` now takes a pessimistic lock on the profile row (`SELECT … FOR UPDATE`) while assigning the next version number, so concurrent edits serialize and each produces a distinct version.
- New unique constraint `uq_notification_profile_version` on `notification_profile_version (notification_profile_uuid, version)`, declared on the entity and added by a Flyway migration. The migration renumbers already-existing duplicate version numbers chronologically and remaps `pending_notification` version references accordingly.
- Recipient info is resolved before the row lock is taken, keeping the auth service call outside the lock window.
- A violation of the new unique constraint is reported as a validation error (profile was concurrently modified); other integrity violations surface unchanged.
- Description changes are persisted even when the edit does not result in a new version.

### CODE REVIEW

File: src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java [hunk 1]
```diff
@@ -144,19 +147,26 @@ public NotificationProfileDetailDto createNotificationProfile(NotificationProfil
...
+        // the auth service over HTTP and must not extend the lock window.
+        List<NameAndUuidDto> recipients = resolveRecipients(updateRequestDto.getRecipientType(), updateRequestDto.getRecipientUuids());
 
-        if (areVersionsEqual(currentVersion, updateRequestDto)) {
-            logger.debug("Current version of notification profile {} is same as in request. New version is not created", notificationProfile.getName());
-            return getNotificationProfileDetailDto(currentVersion);
-        }
+        // The row lock serializes concurrent edits; without it, both could read the same latest version and
+        // insert duplicate version numbers.
+        NotificationProfile notificationProfile = notificationProfileRepository.findAndLockByUuid(uuid.getValue()).orElseThrow(() -> new NotFoundException(NotificationProfile.class, uuid));
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java:183]: Restructured in a83d3dfe6: `editNotificationProfile` now runs with `@Transactional(propagation = NOT_SUPPORTED)`, resolves recipients without a transaction, and delegates the locked version assignment to `persistEditedVersion`, a `@Transactional` method invoked through the lazily self-injected proxy (same pattern as `SigningProfileServiceImpl.persistUpdate`). `getRecipientObjectInfo` additionally suppresses the class-level transaction of its own service, so the auth service HTTP call runs with no DB transaction at all.

File: src/main/resources/db/migration/V202607211000__notification_profile_version_unique_version.sql [hunk 2]
```diff
@@ -0,0 +1,62 @@
+-- Flyway runs this file in a single transaction; the explicit locks block concurrent writers (e.g. the
```
Comment (lubomirw) [inline src/main/resources/db/migration/V202607211000__notification_profile_version_unique_version.sql:1]: **[Medium]** | Testing

The Flyway repair migration's data-repair SQL (version renumbering + `pending_notification` remap) never runs in CI: the integration suite uses `spring.flyway.enabled=false` and `hibernate.ddl-auto=create-drop` (`src/test/resources/application.yml`), so the unique constraint the new tests exercise is generated by Hibernate from the `@UniqueConstraint` entity annotation — not by this migration. The riskiest part of the change is entirely unverified.

_Suggested:_ Add a migration integration test (the repo already has `integration/migration/BaseMigrationTest`) that seeds duplicate `(notification_profile_uuid, version)` rows plus referencing `pending_notification` rows, runs this migration with Flyway enabled, and asserts the dense renumbering and the `pending_notification` remap — including the ambiguous duplicate-version (MAX/latest-created) case.

<sub>flagged by superpowers code-review + Copilot CLI</sub>

File: src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java [hunk 3]
```diff
@@ -167,22 +192,44 @@ public NotificationProfileDetailDto editNotificationProfile(SecuredUUID uuid, No
...
         notificationProfileVersion.setRepetitions(updateRequestDto.getRepetitions());
-        notificationProfileVersion = notificationProfileVersionRepository.save(notificationProfileVersion);
+        try {
+            notificationProfileVersion = notificationProfileVersionRepository.saveAndFlush(notificationProfileVersion);
+        } catch (DataIntegrityViolationException e) {
+            // Backstop for the unique (notification_profile_uuid, version) constraint, reachable only if a
+            // writer bypasses the row lock above. Other integrity violations (e.g. a foreign key on a
+            // concurrently deleted notification instance) are not concurrent-edit collisions and must surface as-is.
+            if (isUniqueVersionViolation(e)) {
+                throw new ValidationException("Notification profile %s was concurrently modified. Retry the edit.".formatted(notificationProfile.getName()));
```
Comment (lubomirw) [inline src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java:214]: **[Low]** | Testing

The branch mapping a duplicate-version `DataIntegrityViolationException` to a `ValidationException` (`isUniqueVersionViolation == true`) is never exercised: the row lock prevents duplicates in the concurrency test, `testEditWithUnknownNotificationInstanceFailsWithIntegrityError` covers only the rethrow branch, and `testDuplicateVersionInsertIsRejectedByUniqueConstraint` only pins the constraint name at the repository level.

_Suggested:_ Unit-test the mapping directly — synthesize a `DataIntegrityViolationException` wrapping a Hibernate `ConstraintViolationException` whose `constraintName` is `NotificationProfileVersion.UNIQUE_VERSION_CONSTRAINT`, assert `persistEditedVersion` surfaces `ValidationException`; add a negative case (different constraint name → rethrow).

<sub>flagged by guided-pr-review + superpowers code-review</sub>

File: src/test/java/com/otilm/core/integration/service/NotificationProfileServiceITest.java [hunk 4]
```diff
@@ -221,6 +236,81 @@ void testUpdateNotificationProfile() throws NotFoundException {
         mockServer.stop();
     }
 
+    @Test
+    void testConcurrentEditsAssignDistinctVersions() throws Exception {
```
Comment (lubomirw) [inline src/test/java/com/otilm/core/integration/service/NotificationProfileServiceITest.java:245]: **[Low]** | Testing

`testConcurrentEditsAssignDistinctVersions` synchronizes only the start (single `CyclicBarrier`, 2 threads, 1 iteration); the read-latest-version-then-insert contention window is tiny, so the test passes deterministically because the lock works — but would frequently still pass by chance even if the pessimistic lock were removed, giving weak regression protection for the feature's core mechanism.

_Suggested:_ Loop the scenario over many iterations and/or use more editor threads, or hold a latch inside `persistEditedVersion` during the test, to force the interleaving the lock is meant to serialize.

<sub>flagged by superpowers code-review</sub>

File: src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java [hunk 5]
```diff
@@ -167,22 +192,44 @@ public NotificationProfileDetailDto editNotificationProfile(SecuredUUID uuid, No
...
+            }
+            throw e;
```
Comment (lubomirw) [inline src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java:216]: **[Low]** | Error handling

A client-supplied unknown notification-instance reference produces a foreign-key `DataIntegrityViolationException` that is rethrown as-is here and then caught by the generic `@ExceptionHandler(Exception.class)`, yielding HTTP 500 for what is really invalid input (no data leak — the generic handler returns a fixed message). Pre-existing behavior, now locked in by `testEditWithUnknownNotificationInstanceFailsWithIntegrityError`.

_Suggested:_ Validate the notification-instance reference before insert (or map the FK violation to a 4xx domain exception) so bad input returns 404/422 instead of 500 — kept distinct from the concurrent-modification mapping above.

<sub>flagged by superpowers code-review</sub>

File: src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java [hunk 6]
```diff
@@ -143,20 +155,33 @@ public NotificationProfileDetailDto createNotificationProfile(NotificationProfil
...
-        NotificationProfileVersion currentVersion = notificationProfile.getCurrentVersion();
+        // Resolve recipient info from the request before opening the write transaction: recipient lookup
+        // can call the auth service over HTTP and must not hold a DB connection or the profile row lock.
+        List<NameAndUuidDto> recipients = resolveRecipients(updateRequestDto.getRecipientType(), updateRequestDto.getRecipientUuids());
 
-        if (areVersionsEqual(currentVersion, updateRequestDto)) {
-            logger.debug("Current version of notification profile {} is same as in request. New version is not created", notificationProfile.getName());
-            return getNotificationProfileDetailDto(currentVersion);
-        }
+        return self.persistEditedVersion(uuid, updateRequestDto, recipients);
```
Comment (lubomirw) [inline src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java:176]: **[Nit]** | Maintainability

`persistEditedVersion`'s `@Transactional` (which scopes the `PESSIMISTIC_WRITE` lock) takes effect because it is invoked through the injected `self` proxy. A later refactor to a direct `this.persistEditedVersion(...)` call would drop the transaction and reintroduce the version race. (Package-private is fine — Spring 6.2 + CGLIB advises it.)

_Suggested:_ Add a one-line comment on this `self.persistEditedVersion(...)` call noting the transaction boundary relies on the self-proxied invocation, so the indirection isn't "simplified away" in a future edit.

<sub>flagged by guided-pr-review</sub>

**Review by lubomirw — COMMENTED** [review]
Automated multi-reviewer triage — 5 findings (COMMENT). Author-selected from a Copilot CLI + guided-pr-review + superpowers + hygiene run; deduped and independently verified.

A separately-claimed **High** ("package-private `@Transactional` gets no advice → lock not held") was **refuted**: Spring Framework 6.2 advises package-visible methods on CGLIB class-based proxies by default, this project sets no `publicMethodsOnly=true` override, and the method is invoked via the `self` proxy — so the `PESSIMISTIC_WRITE` lock is held across the read-modify-write as intended.

File: src/main/resources/db/migration/V202607211000__notification_profile_version_unique_version.sql (hunk 2 above)
Comment (3keyroman) [reply src/main/resources/db/migration/V202607211000__notification_profile_version_unique_version.sql:1]: Added `NotificationProfileVersionUniqueMigrationITest` in 33d5869f5: it drops the entity-generated constraint, seeds duplicate `(notification_profile_uuid, version)` rows with controlled `created_at` plus referencing `pending_notification` rows, executes the migration SQL file in a single transaction (mirroring Flyway), and asserts the dense renumbering, the remap — including the ambiguous duplicated-version → latest-created case — the untouched clean profile, and that the restored constraint rejects duplicates.

File: src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java (hunk 3 above)
Comment (3keyroman) [reply src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java:214]: Added `NotificationProfileServiceImplPersistEditTest` in 33d5869f5: it drives `persistEditedVersion` with synthesized `DataIntegrityViolationException`s — matching constraint name maps to `ValidationException` with the retry message; a different constraint name and a null constraint name are rethrown unchanged (asserted same instance).

File: src/test/java/com/otilm/core/integration/service/NotificationProfileServiceITest.java (hunk 4 above)
Comment (3keyroman) [reply src/test/java/com/otilm/core/integration/service/NotificationProfileServiceITest.java:245]: Strengthened in 33d5869f5: four editors over three rounds, each round on a fresh profile, asserting the exact dense sequence `[1..5]` per round. I left out the latch-inside-`persistEditedVersion` option — it would add test-only state to production code; the widened schedule fails reliably without the lock (the original two-thread variant is what reproduced the duplicate before the fix).

File: src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java (hunk 5 above)
Comment (3keyroman) [reply src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java:216]: Fixed in 33d5869f5: `createNotificationProfile` and `editNotificationProfile` validate the notification instance reference up front and throw `NotFoundException` for an unknown reference, so clients get 404 instead of 500. The foreign key stays as the backstop for a concurrent instance deletion (that rethrow branch is covered by the new unit test). The former integrity-error test now asserts not-found on both paths, including that a failed create leaves no partial profile behind.

File: src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java (hunk 6 above)
Comment (3keyroman) [reply src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java:176]: Added in 33d5869f5 — the call site now notes that the transaction boundary relies on the self-proxied invocation.

File: src/main/java/com/otilm/core/service/impl/ResourceObjectAssociationServiceImpl.java [hunk 7]
```diff
@@ -179,6 +180,9 @@ public void bulkRemoveObjectAssociations(Resource resource, List<UUID> objectUui
     }
 
     @Override
+    // The USER and ROLE branches call the auth service over HTTP; suppressing the class-level
+    // @Transactional keeps transaction-less callers from opening a DB transaction around that call.
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/main/java/com/otilm/core/service/impl/ResourceObjectAssociationServiceImpl.java]: Reworded in fc25a3d94 — the comment now states that a caller's active transaction is suspended for the duration of the call rather than joined.

File: src/test/java/com/otilm/core/integration/service/NotificationProfileServiceITest.java [hunk 8]
```diff
@@ -221,6 +241,102 @@ void testUpdateNotificationProfile() throws NotFoundException {
...
+                        .filter(version -> version.getNotificationProfileUuid().equals(profileUuid.getValue()))
+                        .map(NotificationProfileVersion::getVersion)
+                        .sorted()
+                        .toList();
+                Assertions.assertEquals(expectedVersions, versions,
+                        "Concurrent edits must serialize version assignment and never produce duplicate version numbers (round " + round + ")");
+            }
+        } finally {
+            executor.shutdownNow();
+        }
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/test/java/com/otilm/core/integration/service/NotificationProfileServiceITest.java:289]: Changed in fc25a3d94 to orderly `shutdown()` + `awaitTermination(30s)` with `shutdownNow()` as the timeout fallback.

File: src/test/java/com/otilm/core/integration/migration/NotificationProfileVersionUniqueMigrationITest.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+            // The unique constraint is in place again and rejects duplicates.
+            try (Statement statement = connection.createStatement()) {
+                SQLException rejection = Assertions.assertThrows(SQLException.class, () -> statement.execute("""
+                        INSERT INTO notification_profile_version
+                            (uuid, notification_profile_uuid, version, recipient_type, internal_notification, created_at)
+                        VALUES ('c0000000-0000-0000-0000-000000000001', '%s', 1, 'OWNER', true, now())
+                        """.formatted(PROFILE_A)));
+                Assertions.assertTrue(rejection.getMessage().contains("uq_notification_profile_version"),
+                        "Duplicate insert should be rejected by the unique constraint, but was: " + rejection.getMessage());
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/test/java/com/otilm/core/integration/migration/NotificationProfileVersionUniqueMigrationITest.java:81]: Added in fc25a3d94 — the test now asserts SQLSTATE 23505 alongside the constraint name (the name check stays to pin *which* constraint fired).

File: src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java [hunk 10]
```diff
@@ -143,20 +164,36 @@ public NotificationProfileDetailDto createNotificationProfile(NotificationProfil
...
     @Override
     @ExternalAuthorization(resource = Resource.NOTIFICATION_PROFILE, action = ResourceAction.UPDATE)
+    @Transactional(propagation = Propagation.NOT_SUPPORTED)
     public NotificationProfileDetailDto editNotificationProfile(SecuredUUID uuid, NotificationProfileUpdateRequestDto updateRequestDto) throws NotFoundException {
-        NotificationProfile notificationProfile = notificationProfileRepository.findByUuid(uuid).orElseThrow(() -> new NotFoundException(NotificationProfile.class, uuid));
-        NotificationProfileVersion currentVersion = notificationProfile.getCurrentVersion();
+        // Resolve recipient info from the request before opening the write transaction: recipient lookup
+        // can call the auth service over HTTP and must not hold a DB connection or the profile row lock.
+        List<NameAndUuidDto> recipients = resolveRecipients(updateRequestDto.getRecipientType(), updateRequestDto.getRecipientUuids());
+        validateNotificationInstanceExists(updateRequestDto.getNotificationInstanceUuid());
```
[the remark this answers is not in the document]
Comment (3keyroman) [reply src/main/java/com/otilm/core/service/impl/NotificationProfileServiceImpl.java]: Reordered in the follow-up commit — the instance existence check now runs before recipient resolution, so an invalid instance UUID fails fast without any auth service call.

**Review by lubomirw — APPROVED** [review]
